drivers/serial/uart_tt_virt.c

```diff
@@ -0,0 +1,418 @@
+		}
+	}
+
+	if (config->loopback && size > 0) {
+		K_SPINLOCK(&data->rx_lock) {
+			int lim = MIN(size, (int)tt_vuart_buf_space(vuart->rx_head, vuart->rx_tail,
+								    vuart->rx_cap));
+
+			for (int i = 0; i < lim; ++i) {
+				unsigned char ch = -1;
+
+				(void)tt_vuart_poll_in(vuart, &ch, TT_VUART_ROLE_HOST);
+				tt_vuart_poll_out(vuart, ch, TT_VUART_ROLE_HOST);
+			}
+
+			/* Note: irq_handler() picks up rx data */
+		}
+	}
+
+	return size;
+}
+
+static int uart_tt_virt_fifo_read(const struct device *dev, uint8_t *rx_data, int size)
+{
+	struct uart_tt_virt_data *data = dev->data;
+	const struct uart_tt_virt_config *config = dev->config;
+	volatile struct tt_vuart *vuart = config->vuart;
+
+	__ASSERT_NO_MSG(size >= 0);
+
+	K_SPINLOCK(&data->rx_lock) {
+		size = MIN(size, (int)tt_vuart_buf_size(vuart->rx_head, vuart->rx_tail));
+
+		for (int i = 0; i < size; ++i) {
+			(void)tt_vuart_poll_in(vuart, rx_data++, TT_VUART_ROLE_DEVICE);
+		}
+	}
+
+	return size;
+}
+
+static void uart_tt_virt_irq_callback_set(const struct device *dev,
+					  uart_irq_callback_user_data_t cb, void *user_data)
+{
+	struct uart_tt_virt_data *const data = dev->data;
+
+	data->irq_cb = cb;
+	data->irq_cb_udata = user_data;
+}
+
+static void uart_tt_virt_irq_err_disable(const struct device *dev)
+{
+	struct uart_tt_virt_data *const data = dev->data;
+
+	K_SPINLOCK(&data->err_lock) {
+		data->err_irq_en = false;
+	}
+}
+
+static void uart_tt_virt_irq_err_enable(const struct device *dev)
+{
+	bool submit;
+	struct uart_tt_virt_data *const data = dev->data;
+
+	K_SPINLOCK(&data->err_lock) {
+		data->err_irq_en = true;
+		submit = !!data->err_flags;
+	}
+
+	if (submit) {
+		(void)k_work_submit(&data->irq_work);
+	}
+}
+
+static void uart_tt_virt_irq_handler(struct k_work *work)
+{
+	struct uart_tt_virt_data *data = CONTAINER_OF(work, struct uart_tt_virt_data, irq_work);
+	const struct device *dev = data->dev;
+	uart_irq_callback_user_data_t cb = data->irq_cb;
+	void *udata = data->irq_cb_udata;
+
+	if (cb == NULL) {
+		LOG_DBG("No IRQ callback configured for uart_tt_virt device %p", dev);
+		return;
+	}
+
+	while (uart_tt_virt_irq_is_pending(dev)) {
+		cb(dev, udata);
+	}
+}
+
+static int uart_tt_virt_irq_is_pending(const struct device *dev)
+{
+	return uart_tt_virt_irq_tx_ready(dev) || uart_tt_virt_irq_rx_ready(dev);
+}
+
+static void uart_tt_virt_irq_rx_disable(const struct device *dev)
+{
+	struct uart_tt_virt_data *const data = dev->data;
+
+	K_SPINLOCK(&data->rx_lock) {
+		data->rx_irq_en = false;
+	}
+}
+
+static void uart_tt_virt_irq_rx_enable(const struct device *dev)
+{
+	bool submit;
+	const struct uart_tt_virt_config *config = dev->config;
+	struct uart_tt_virt_data *const data = dev->data;
+	volatile struct tt_vuart *vuart = config->vuart;
+
+	K_SPINLOCK(&data->rx_lock) {
+		data->rx_irq_en = true;
+		submit = !tt_vuart_buf_empty(vuart->rx_head, vuart->rx_tail);
+	}
+
+	if (submit) {
+		(void)k_work_submit(&data->irq_work);
```

Comment:
Won't this only run the IRQ worker once, and then stop? Or is there somewhere where we reschedule the IRQ worker to poll periodically?

Reply:
It's only intended to run the IRQ worker once.